### PR TITLE
keep and follower counts for /ext/page libraries

### PR DESCRIPTION
**Pull request by 2is10**
and excluding libraries that the user already follows or collaborates on (in addition to those he/she owns)

also removing `keepersOmitted` from the response because we aren’t planning to use the total count of friends who’ve kept anywhere in the UI

@eishay

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMembershipRepo.scala [hunk 1]
```diff
@@ -120,10 +124,13 @@ class LibraryMembershipRepoImpl @Inject() (
...
   def getWithLibraryIdsAndUserId(libraryIds: Set[Id[Library]], userId: Id[User], excludeState: Option[State[LibraryMembership]] = Some(LibraryMembershipStates.INACTIVE))(implicit session: RSession): Map[Id[Library], LibraryMembership] = {
-    if (libraryIds.isEmpty) {
-      Map.empty
-    } else {
-      (for (b <- rows if b.libraryId.inSet(libraryIds) && b.userId === userId && b.state =!= excludeState.orNull) yield (b.libraryId, b)).list.toMap
+    libraryIds.size match {
+      case 0 =>
+        Map.empty
+      case 1 =>
+        getWithLibraryIdAndUserId(libraryIds.head, userId, excludeState) map { lm => (lm.libraryId -> lm) } toMap
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMembershipRepo.scala:131]: an optimization: using compiled query in the `1` case

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMembershipRepo.scala [hunk 2]
```diff
@@ -161,6 +175,35 @@ class LibraryMembershipRepoImpl @Inject() (
...
+      case 1 =>
+        val libraryId = libraryIds.head
+        Map(libraryId -> countWithLibraryIdAndAccess(libraryId, access))
+      case n =>
+        val ids = Seq.fill(n)("?").mkString(",")
+        val stmt = session.getPreparedStatement(s"select library_id, count(*) from library_membership where library_id in ($ids) and access='${access.value}' and state='active' group by library_id")
+        libraryIds.zipWithIndex.foreach {
+          case (id, idx) =>
+            stmt.setLong(idx + 1, id.id)
+        }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMembershipRepo.scala:192]: I copied this prepared statement pattern from elsewhere in our codebase. I don’t know whether it’s the best pattern or not. It’s sure a lot of code.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PageCommander.scala [hunk 3]
```diff
@@ -129,30 +129,37 @@ class PageCommander @Inject() (
...
-              Json.obj(
-                "name" -> library.name,
-                "slug" -> library.slug,
-                "owner" -> basicUserMap(userId),
-                "keeps" -> numKeeps,
-                "followers" -> numFollowers)
+          val otherKeepersTotal = info.keepersTotal - (if (userIdSet.contains(userId)) 1 else 0)
+          val (keepCounts, followerCounts) = db.readOnlyReplica { implicit session =>
+            val libraryIds = libraries.map(_.id.get).toSet
+            val keepCounts = keepRepo.getCountsByLibrary(libraryIds)
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PageCommander.scala]: `getCountsByLibrary` is cached, the cache will not be used

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PageCommander.scala [hunk 4]
```diff
@@ -129,30 +129,37 @@ class PageCommander @Inject() (
...
+            val followerCounts = libraryMembershipRepo.countWithAccessByLibraryId(libraryIds, LibraryAccess.READ_ONLY)
+            (keepCounts, followerCounts)
+          }
+          val libraryObjs = libraries.map { lib => // TODO: sort by friends first, secondarily by num followers (or just trust search ordering?)
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PageCommander.scala]: Lets see how search ordering goes before investing too much in it, not sure how it sorts.

**Comment by eishay** [comment]
cool, thanks!

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PageCommander.scala (hunk 3 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PageCommander.scala]: moved it up to the `readOnlyMaster` session. thanks

**Pull request merged**